bson-codec-design/bson/decoder.go

```diff
@@ -0,0 +1,23 @@
+package bson
+
+import "io"
+
+// A Decoder reads and decodes BSON documents from a stream.
+type Decoder struct{}
+
+// NewDecoder returns a new decoder that uses Registry reg to read serialization type st from r.
+func NewDecoder(reg *Registry, r io.Reader, st SerializationType) (*Decoder, error) { return nil, nil }
```

Comment:
1. Not sure why we need a reader here. That should be subsumed by the ValueReader that the user is using (see below comment). The problem is that a reader isn't necessary in all circumstances, for instance, when going to a Document.

2. I think the serialization type is weird here. The user should just be able to pass us an implementation of a ValueReader. We'll have a ExtJson version, a Bson version, a Document version... and the user can create their own if they want.

In addition, you could have helper methods for our built in ones... NewDocumentDecoder, NewExtJsonDecoder, etc...

The API should look something like this:

```
// For bson, r = io.Reader and registry = Registry
bsonReader := NewBsonValueReader(r)
decoder := NewDecoder(registry, bsonReader)

// short version
decoder:= NewBsonDecoder(registry, r)
```

```
// For documents, doc = bson.Document and registry = Registry
docReader := NewDocumentValueReader(doc)
decoder:= NewDecoder(registry, docReader)

// short version
decoder:= NewDocumentDecoder(registry, doc)
```

Reply:
I don't particularly like exposing the ValueReader and ValueWriter implementation types, it's really complicating the package concerning naming. I also find the extra step of wrapping an io.Reader or io.Writer in a ValueReader or ValueWriter to be of little value, you can't do anything with it except for use with a Decoder or Encoder (at least I don't think we want people to be arbitrarily using these types).

We could remove the SerializationType and just use `NewBsonDecoder` and `NewExtJSONDecoder` methods. Overall, this package feels like it's continually growing with exposed parts that require people to carefully assemble things. This isn't the Core Library, and I don't think verbosity works well here, especially for people coming from `mgo/bson`.